### PR TITLE
[JBTM-3739] Create Narayana BOM

### PR DESCRIPTION
**Pull request by moulalis**
Issue: https://issues.redhat.com/browse/JBTM-3739

CORE TOMCAT AS_TESTS RTS JACOCO XTS QA_JTA QA_JTS_OPENJDKORB LRA

Skipped
!PERF !DB_TESTS !mysql !db2 !postgres !oracle

### CODE REVIEW

**Comment by jmfinelli** [comment]
Thanks for your PR @moulalis! :tada:

**Comment by marcosgopen** [comment]
TESTIT

File: narayana-bom/pom.xml [hunk 1]
```diff
@@ -0,0 +1,1066 @@
...
+          <exclusion>
+            <groupId>org.jboss.logging</groupId>
+            <artifactId>jboss-logging-spi</artifactId>
+          </exclusion>
+        </exclusions>
+      </dependency>
+      <dependency>
+        <groupId>org.springframework</groupId>
+        <artifactId>spring-jdbc</artifactId>
+        <version>3.2.2.RELEASE</version>
```
Comment (marcosgopen) [inline narayana-bom/pom.xml:196]: Could I ask you to set the version with a property? Thanks

File: narayana-bom/pom.xml [hunk 2]
```diff
@@ -0,0 +1,1066 @@
...
+      <dependency>
+        <groupId>mysql</groupId>
+        <artifactId>mysql-connector-java</artifactId>
+        <version>${version.mysql}</version>
+        <scope>test</scope>
+      </dependency>
+      <dependency>
+        <groupId>jboss</groupId>
+        <artifactId>jnpserver</artifactId>
+        <version>4.2.2.GA</version>
```
Comment (marcosgopen) [inline narayana-bom/pom.xml:273]: Could I ask you to set the version with a property? Thanks

**Review by marcosgopen — COMMENTED** [review]
Thank you very much @moulalis for you PR! I wrote some little comments but generally it looks great. I think it will be very useful to have the bom in Narayana. And also other Narayana projects could then use it!

**Comment by mmusgrov** [comment]
I was expecting that each narayana module pom use the bom, otherwise the versions used by narayana and narayana consumers will diverge. So each narayana module pom should contain: 

```
  <dependencyManagement>
    <dependencies>
      <dependency>
        <groupId>org.jboss.narayana</groupId>
        <artifactId>narayana-bom</artifactId>
        <version>6.0.0.Final-SNAPSHOT</version>
        <type>pom</type>
        <scope>import</scope>
      </dependency>
    </dependencies>
  </dependencyManagement>
```

**Comment by jmfinelli** [comment]
> I was expecting that each narayana module pom use the bom, otherwise the versions used by narayana and narayana consumers will diverge. So each narayana module pom should contain:
> 
> ```
>   <dependencyManagement>
>     <dependencies>
>       <dependency>
>         <groupId>org.jboss.narayana</groupId>
>         <artifactId>narayana-bom</artifactId>
>         <version>6.0.0.Final-SNAPSHOT</version>
>         <type>pom</type>
>         <scope>import</scope>
>       </dependency>
>     </dependencies>
>   </dependencyManagement>
> ```

I guess @moulalis is planning to create a [follow up](https://issues.redhat.com/browse/JBTM-3740) PR

**Comment by moulalis** [comment]
> > I was expecting that each narayana module pom use the bom, otherwise the versions used by narayana and narayana consumers will diverge. So each narayana module pom should contain:
> > ```
> >   <dependencyManagement>
> >     <dependencies>
> >       <dependency>
> >         <groupId>org.jboss.narayana</groupId>
> >         <artifactId>narayana-bom</artifactId>
> >         <version>6.0.0.Final-SNAPSHOT</version>
> >         <type>pom</type>
> >         <scope>import</scope>
> >       </dependency>
> >     </dependencies>
> >   </dependencyManagement>
> > ```
> 
> I guess @moulalis is planning to create a [follow up](https://issues.redhat.com/browse/JBTM-3740) PR

Yes, @mmusgrov I've created a subtask as https://issues.redhat.com/browse/JBTM-3740, will send a separate PR for the same.

File: narayana-bom/pom.xml (hunk 1 above)
Comment (moulalis) [reply narayana-bom/pom.xml]: This is updated.

File: narayana-bom/pom.xml (hunk 2 above)
Comment (moulalis) [reply narayana-bom/pom.xml]: This is updated.

**Comment by mmusgrov** [comment]
The two PRs would depend on each other. If we merge this one and find something is missing when we do the sub-task then we'd need to reopen this task/PR. Can you include the work for both tasks in this PR please.

**Comment by mmusgrov** [comment]
It's looking very good indeed. I did notice that some dependencies still specify the version (bypassing the bom), for example artifact "org.jboss:jboss-transaction-spi", "org.jboss.openjdk-orb:openjdk-orb" etc

**Comment by mmusgrov** [comment]
When this one is eventually merged we need to use it in the [quickstart repo](https://github.com/jbosstm/quickstart) or even testing the quickstart repo now is a possibility.

File: ArjunaJTA/jta/pom.xml [hunk 3]
```diff
@@ -272,13 +251,16 @@
         <dependency>
           <groupId>org.wildfly.arquillian</groupId>
           <artifactId>wildfly-arquillian-container-managed</artifactId>
-          <version>${version.org.wildfly.arquillian}</version>
           <scope>test</scope>
         </dependency>
         <dependency>
           <groupId>org.jboss</groupId>
+<<<<<<< HEAD
```
Comment (mmusgrov) [inline ArjunaJTA/jta/pom.xml:258]: @moulalis will you resolve this merge conflict please.

File: narayana-full/pom.xml [hunk 4]
```diff
@@ -133,9 +128,16 @@
       <scope>provided</scope>
     </dependency>
     <dependency>
+<<<<<<< HEAD
```
Comment (mmusgrov) [inline narayana-full/pom.xml:131]: And here too.

**Review by mmusgrov — DISMISSED** [review]

File: txbridge/pom.xml [hunk 5]
```diff
@@ -187,159 +191,153 @@
...
         <dependency>
             <groupId>org.glassfish</groupId>
             <artifactId>jakarta.json</artifactId>
-            <version>${version.glassfish.json}</version>
             <scope>test</scope>
         </dependency>
 
         <dependency>
-            <groupId>org.jboss</groupId>
+            <groupId>org.jboss</groupId>S
```
Comment (mmusgrov) [inline txbridge/pom.xml:295]: Spurious character at end of line.

File: ArjunaJTA/jta/pom.xml (hunk 3 above)
Comment (moulalis) [reply ArjunaJTA/jta/pom.xml]: > @moulalis will you resolve this merge conflict please.

@mmusgrov Good catch, I've updated the changes.

File: narayana-full/pom.xml (hunk 4 above)
Comment (moulalis) [reply narayana-full/pom.xml]: @mmusgrov I've updated the changes.

File: txbridge/pom.xml (hunk 5 above)
Comment (moulalis) [reply txbridge/pom.xml]: @mmusgrov I've updated the changes.

File: pom.xml [hunk 6]
```diff
@@ -190,12 +191,15 @@
...
     <version.com.microsoft.sqlserver>mssql2005_sqljdbc_2.0</version.com.microsoft.sqlserver>
     <version.com.oracle>18.3.0.0</version.com.oracle>
+    <version.com.sun.messaging.mq.fscontext>4.6-b01</version.com.sun.messaging.mq.fscontext>
     <version.com.sybase.jConnect>6.0</version.com.sybase.jConnect>
     <version.commons-httpclient>3.1-jbossorg-1</version.commons-httpclient>
     <version.compiler.plugin>3.8.0-jboss-2</version.compiler.plugin>
     <version.doxia>1.8.1</version.doxia>
     <version.glassfish.json>2.0.1</version.glassfish.json>
     <version.hamcrest>2.2</version.hamcrest>
+    <version.hamcrest>1.3</version.hamcrest>
```
Comment (marcosgopen) [inline pom.xml:201]: Thanks @moulalis for your PR! Please use only one property for version.hamcrest (lines 200 and 201)

Comment (marcosgopen) [reply pom.xml:201]: The dependency hamcrest-library (version 1.3) was moved to hamcrest dependency (https://mvnrepository.com/artifact/org.hamcrest/hamcrest-library). I think we should use version 2.2, Thanks

File: narayana-bom/pom.xml [hunk 7]
```diff
@@ -0,0 +1,1107 @@
...
+
+      <dependency>
+        <groupId>org.hamcrest</groupId>
+        <artifactId>hamcrest</artifactId>
+        <version>${version.hamcrest}</version>
+        <scope>test</scope>
+      </dependency>
+      <dependency>
+        <groupId>org.hamcrest</groupId>
+        <artifactId>hamcrest-library</artifactId>
```
Comment (marcosgopen) [inline narayana-bom/pom.xml:1082]: I would remove this dependency and keep only the hamcrest dependency above.

File: pom.xml (hunk 6 above)
Comment (marcosgopen) [reply pom.xml:201]: See PR I created https://github.com/jbosstm/narayana/pull/2086

Comment (moulalis) [reply pom.xml]: Thanks @marcosgopen, my bad I missed removing it.

File: narayana-bom/pom.xml (hunk 7 above)
Comment (moulalis) [reply narayana-bom/pom.xml]: Updated the PR.

File: pom.xml (hunk 6 above)
Comment (marcosgopen) [reply pom.xml]: Thanks

**Comment by mmusgrov** [comment]
Hi @moulalis , I notice that you are still working on this PR so I have added the `Hold` label. When you are ready for a review please can you ping one of the team.

**Comment by moulalis** [comment]
> Hi @moulalis , I notice that you are still working on this PR so I have added the `Hold` label. When you are ready for a review please can you ping one of the team.

@mmusgrov This PR is ready for review, getting conflict on PR after each PR is merged, so resolving it. I just saw there is another [PR](https://github.com/jbosstm/narayana/commit/691a36b2cd2f1ac5f421da397c22b0c832cfcfac) that got merged so here it is showing conflict again. I have to work on it now. :)

**Comment by mmusgrov** [comment]
Sorry about that; your change affects the whole build system and the benefit will be worth it!

**Comment by jmfinelli** [comment]
Kicked off (nearly) all testing axis. Thanks @moulalis

File: README.md [hunk 8]
```diff
@@ -79,6 +79,22 @@ STM: `./build.[sh|bat] -am -pl :stm`
 
 LRA: `./build.[sh|bat] -am -f rts/lra`
 
+Narayan (BOM) Bill Of Materials
```
Comment (iweiss) [inline README.md:82]: Narayana*

**Comment by marcosgopen** [comment]
The good news is that Narayana builds fine now, but we have an error building Wildfly: `[ERROR] Failed to execute goal on project wildfly-iiop-openjdk: Could not resolve dependencies for project org.wildfly:wildfly-iiop-openjdk:jar:28.0.0.Beta1-SNAPSHOT: Failed to collect dependencies at org.jboss.narayana.jts:narayana-jts-idlj:jar:6.0.1.Final-SNAPSHOT: Failed to read artifact descriptor for org.jboss.narayana.jts:narayana-jts-idlj:jar:6.0.1.Final-SNAPSHOT: Failure to find org.jboss.narayana:narayana-bom:pom:6.0.1.Final-SNAPSHOT in https://repository.jboss.org/nexus/content/repositories/thirdparty-releases/ was cached in the local repository, resolution will not be reattempted until the update interval of jbossThirdParty has elapsed or updates are forced -> [Help 1]`

**Comment by mmusgrov** [comment]
@moulalis I notice that your PR builds the narayana-bom module after the other modules but it should be the first one otherwise it won't get installed into the local .m2 repo. I did try building the PR and I get:

```
When I build his PR I get:
[mmusgrov@dev2 rest]$ ls ~/.m2/repository/org/jboss/narayana/narayana-bom/6.0.1.Final-SNAPSHOT/
narayana-bom-6.0.1.Final-SNAPSHOT.pom.lastUpdated  resolver-status.properties
so it should definitely be the first module to be built.
```

which would explain why the WildFly build cannot find it.

**Comment by mmusgrov** [comment]
Ah wait, the issue is that it is missing the `maven-install-plugin` plugin so it doesn't install correctly.

**Comment by moulalis** [comment]
Hi @marcosgopen @mmusgrov @jmfinelli,  This PR is ready for review.
Thanks.

**Review by marcosgopen — CHANGES_REQUESTED** [review]
I can see we have the dependencyManagement configured inside all the modules, may we include it only in the root pom instead?

File: ArjunaCore/pom.xml [hunk 9]
```diff
@@ -27,6 +27,19 @@
   <properties>
     <sortpom.skip>true</sortpom.skip>
   </properties>
+
```
[the remark this answers is not in the document]
Comment (marcosgopen) [reply ArjunaCore/pom.xml:30]: I think including the dependencyManagement in the root pom might provoke a maven dependency cycle so I agree to include the sections in the submodules.

**Comment by marcosgopen** [comment]
Thanks @moulalis for you PR, well done! I therefore approve it.
AS_TESTS tests failures are not related to this PR.

**Pull request merged**